In [1]:
import requests
import json
import pandas as pd
import tweepy

In [2]:
with open(r'C:\Users\roast\Desktop\twitter_keys.txt', 'r') as file:
    tokens = file.readlines()
API_KEY = tokens[0]
API_SECRET = tokens[1]
ACCESS_KEY = tokens[2]
ACCESS_SECRET = tokens[3]
bearer = tokens[4]

In [3]:
auth = tweepy.OAuthHandler(API_KEY, API_SECRET)
auth.set_access_token(ACCESS_KEY, ACCESS_SECRET)
client = tweepy.Client(bearer_token = bearer)

In [4]:
def df_response(res):
    tw_users1 = {tw_user['id']: tw_user for tw_user in res.includes['users']}
    df = pd.DataFrame()

    for tweet in res.data:
        if tw_users1[tweet.author_id]:
            tw_user = tw_users1[tweet.author_id]
        temp_dict = {
            'tweet_id': [tweet.id],
            'username': [tw_user.username],
            'text': [tweet.text],
            'created_at': [tweet.created_at],
            'location': [tw_user.location],
            'source': [tweet.source],
            'url': ["https://twitter.com/"+str(tw_user.username)+"/status/"+str(tweet.id)]
        }
        temp_df = pd.DataFrame(temp_dict)
        df = pd.concat([temp_df, df])

    return df

In [5]:
#Creating loop for word: mitaka
mitaka_df = pd.DataFrame()
query1 = '(#mitaka OR mitaka) -is:retweet -is:quote -is:reply lang:en'
last_tw_id = None

while mitaka_df.empty:
    try:
        for i in reversed(range(1,11)):
            tweets1 = client.search_recent_tweets(query=query1, max_results = 100, until_id = last_tw_id, 
                                                user_fields = ['username', 'location'],
                                                tweet_fields=['created_at', 'source'],
                                                expansions = ['author_id'])
            temp_df = df_response(tweets1).sort_values(['created_at'], ascending=False)

            last_tw_id = temp_df.iloc[len(temp_df)-1]['tweet_id']
            
            mitaka_df = pd.concat([temp_df, mitaka_df])
    except KeyError:
        pass

In [6]:
mitaka_df.nunique()

tweet_id      145
username      106
text          138
created_at    144
location       55
source         14
url           145
dtype: int64

In [7]:
mitaka_df.sort_values(['created_at'], ascending=False).iloc[0] #Most Recent Tweet

tweet_id                                    1558923867258818566
username                                          retromanjapan
text          The Showa Retro In #MITAKA #Tokyo #28: Marks O...
created_at                            2022-08-14 21:09:48+00:00
location                                           Tokyo, Japan
source                                        Revive Social App
url           https://twitter.com/retromanjapan/status/15589...
Name: 0, dtype: object

In [8]:
mitaka_df.sort_values(['created_at'], ascending=False).iloc[len(mitaka_df)-1] #Last tweet queried by API allowable

tweet_id                                    1556444156075339776
username                                                hyegumi
text                              LRT MITAKA'S GETTING ANIMATED
created_at                            2022-08-08 00:56:18+00:00
location                                                she/her
source                                      Twitter for Android
url           https://twitter.com/hyegumi/status/15564441560...
Name: 0, dtype: object

In [9]:
mitaka_df.head()

tweet_id       username  \
0  1556927525313323009       denpochi   
0  1556917447805898754       Lizsa___   
0  1556905368520998913  politicfemboy   
0  1556888261964214272    killmekirby   
0  1556884584977006592  retromanjapan   

                                                text  \
0  hey, could u stop posting about mitaka? i ship...   
0  join friend's style challenge\nAsa Mitaka from...   
0  I desire the same as Asa Mitaka... https://t.c...   
0  why does my asa mitaka drawing look like yumek...   
0  The Showa Retro In #MITAKA #Tokyo #8: A One-St...   

                 created_at                                         location  \
0 2022-08-09 08:57:03+00:00  17 they/them\ncsm bllk jjk opm\nic kamonegioisi   
0 2022-08-09 08:17:00+00:00                                             None   
0 2022-08-09 07:29:00+00:00                                             None   
0 2022-08-09 06:21:01+00:00                                     she/her • 17   
0 2022-08-09 06:06:25+00:00                                     Tokyo, Japan   

                source                                                url  
0  Twitter for Android  https://twitter.com/denpochi/status/1556927525...  
0  Twitter for Android  https://twitter.com/Lizsa___/status/1556917447...  
0      Twitter Web App  https://twitter.com/politicfemboy/status/15569...  
0   Twitter for iPhone  https://twitter.com/killmekirby/status/1556888...  
0    Revive Social App  https://twitter.com/retromanjapan/status/15568...

In [10]:
#Creating loop for word: asa mitaka
asamitaka_df = pd.DataFrame()
query2 = '(asamitaka OR "asa mitaka" OR #asamitaka) -is:retweet -is:quote -is:reply lang:en'
last_tw_id = None

while asamitaka_df.empty:
    try:
        for i in reversed(range(1,11)):
            tweets2 = client.search_recent_tweets(query=query2, max_results = 100, until_id = last_tw_id,
                                            user_fields = ['username', 'location'],
                                            tweet_fields=['created_at', 'source'],
                                            expansions = ['author_id'])
            temp_df = df_response(tweets2).sort_values(['created_at'], ascending=False)
            last_tw_id = temp_df.iloc[len(temp_df)-1]['tweet_id']
                
            asamitaka_df = pd.concat([temp_df, asamitaka_df])
    except KeyError:
        pass

In [11]:
asamitaka_df.shape

(90, 7)

In [12]:
asamitaka_df.sort_values(['created_at'], ascending=False).iloc[0]

tweet_id                                    1558850388824150019
username                                        spookygraveyard
text          Do y’all think Asa Mitaka killing Bucky will b...
created_at                            2022-08-14 16:17:49+00:00
location                            it/haunts/envyself lesbian 
source                                       Twitter for iPhone
url           https://twitter.com/spookygraveyard/status/155...
Name: 0, dtype: object

In [13]:
#Last tweet queried by API allowable
asamitaka_df.sort_values(['created_at'], ascending=False).iloc[len(asamitaka_df)-1] 

tweet_id                                    1556457835705221120
username                                              yukizhien
text          I HAV #asamitaka ART LATUUURRRR AAAAAAAA https...
created_at                            2022-08-08 01:50:40+00:00
location                                                     ph
source                                      Twitter for Android
url           https://twitter.com/yukizhien/status/155645783...
Name: 0, dtype: object

In [14]:
asamitaka_df.head()

tweet_id         username  \
0  1558850388824150019  spookygraveyard   
0  1558799546750160896        najinteru   
0  1558654350645514240         cyetelie   
0  1558627510732169218      rrrgphoenix   
0  1558618247318183938       SolwayDemi   

                                                text  \
0  Do y’all think Asa Mitaka killing Bucky will b...   
0  Hi moots!! I missed u all!! Sorry for the long...   
0  new drawing  #asamitaka #chainsawmanfanart htt...   
0  Chainsaw Man live action cast 🤔\n\ntom holland...   
0  updated kinlist yet again\nwas considering put...   

                 created_at                     location               source  \
0 2022-08-14 16:17:49+00:00  it/haunts/envyself lesbian    Twitter for iPhone   
0 2022-08-14 12:55:47+00:00                         None  Twitter for Android   
0 2022-08-14 03:18:50+00:00                         None      Twitter Web App   
0 2022-08-14 01:32:11+00:00                         None      Twitter Web App   
0 2022-08-14 00:55:22+00:00               Forbidden Zone   Twitter for iPhone   

                                                 url  
0  https://twitter.com/spookygraveyard/status/155...  
0  https://twitter.com/najinteru/status/155879954...  
0  https://twitter.com/cyetelie/status/1558654350...  
0  https://twitter.com/rrrgphoenix/status/1558627...  
0  https://twitter.com/SolwayDemi/status/15586182...

In [15]:
#Creating loop for word: mitaka asa
mitakaasa_df = pd.DataFrame()
query3 = '(mitakaasa OR "mitaka asa" OR #mitakaasa) -is:retweet -is:quote -is:reply lang:en'
last_tw_id = None

while mitakaasa_df.empty:
    try:
        for i in reversed(range(1,11)):
            tweets3 = client.search_recent_tweets(query=query3, max_results = 100, until_id = last_tw_id,
                                            user_fields = ['username', 'location'],
                                            tweet_fields=['created_at', 'source'],
                                            expansions = ['author_id'])
            temp_df = df_response(tweets3).sort_values(['created_at'], ascending=False)
            last_tw_id = temp_df.iloc[len(temp_df)-1]['tweet_id']
                
            mitakaasa_df = pd.concat([temp_df, mitakaasa_df])
    except KeyError:
        pass

In [16]:
mitakaasa_df.shape

(13, 7)

In [17]:
mitakaasa_df.sort_values(['created_at'], ascending=False).iloc[0]

tweet_id                                    1558740227119775744
username                                                dnbooru
text          World\n#chainsaw_man\nCharacters\n#mitaka_asa\...
created_at                            2022-08-14 09:00:04+00:00
location                                                  Tokyo
source                                             Danbooru Bot
url           https://twitter.com/dnbooru/status/15587402271...
Name: 0, dtype: object

In [18]:
#Last tweet queried by API allowable
mitakaasa_df.sort_values(['created_at'], ascending=False).iloc[len(mitakaasa_df)-1] 

tweet_id                                    1556756075663089665
username                                        EmAfterMidnight
text          Midnight Rendezvous \n\n#YoshiAsa \n\nNSFW Pri...
created_at                            2022-08-08 21:35:46+00:00
location                                                   None
source                                          Twitter Web App
url           https://twitter.com/EmAfterMidnight/status/155...
Name: 0, dtype: object

In [19]:
mitakaasa_df.head()

tweet_id      username  \
0  1558740227119775744       dnbooru   
0  1558672272600219649       dnbooru   
0  1558386374423719938    todomeowmo   
0  1558237541689872384  NightcoreR45   
0  1558206783289688064     ShockSans   

                                                text  \
0  World\n#chainsaw_man\nCharacters\n#mitaka_asa\...   
0  World\n#chainsaw_man\nCharacters\n#mitaka_asa\...   
0  caught up with chainsaw man abs oh no I think ...   
0  ✦➥  PLEASE HELP ME IN MAKING MY YOUTUBE VIDEO ...   
0  little drawing I made days ago dedicated to Mi...   

                 created_at              location               source  \
0 2022-08-14 09:00:04+00:00                 Tokyo         Danbooru Bot   
0 2022-08-14 04:30:03+00:00                 Tokyo         Danbooru Bot   
0 2022-08-13 09:33:59+00:00  todomomo arc! 🍑🍰 #轟百   Twitter for iPhone   
0 2022-08-12 23:42:35+00:00                  None  Twitter for Android   
0 2022-08-12 21:40:21+00:00                  None  Twitter for Android   

                                                 url  
0  https://twitter.com/dnbooru/status/15587402271...  
0  https://twitter.com/dnbooru/status/15586722726...  
0  https://twitter.com/todomeowmo/status/15583863...  
0  https://twitter.com/NightcoreR45/status/155823...  
0  https://twitter.com/ShockSans/status/155820678...

In [20]:
#Creating loop for word: war devil
wardevil_df = pd.DataFrame()
query4 = '(wardevil OR "war devil" OR #wardevil) -is:retweet -is:quote -is:reply lang:en'
last_tw_id = None

while wardevil_df.empty:
    try:
        for i in reversed(range(1,11)):
            tweets4 = client.search_recent_tweets(query=query4, max_results = 100, until_id = last_tw_id,
                                            user_fields = ['username', 'location'],
                                            tweet_fields=['created_at', 'source'],
                                            expansions = ['author_id'])
            temp_df = df_response(tweets4).sort_values(['created_at'], ascending=False)
            last_tw_id = temp_df.iloc[len(temp_df)-1]['tweet_id']
                
            wardevil_df = pd.concat([temp_df, wardevil_df])
    except KeyError:
        pass

In [21]:
wardevil_df.shape

(195, 7)

In [22]:
wardevil_df.sort_values(['created_at'], ascending=False).iloc[0]

tweet_id                                    1558932756641579009
username                                        xTrillValentine
text          war devil 😈 \n\n#chainsawman https://t.co/nU4m...
created_at                            2022-08-14 21:45:07+00:00
location                                           St Louis, MO
source                                       Twitter for iPhone
url           https://twitter.com/xTrillValentine/status/155...
Name: 0, dtype: object

In [23]:
#Last tweet queried by API allowable
wardevil_df.sort_values(['created_at'], ascending=False).iloc[len(wardevil_df)-1] 

tweet_id                                    1556413306956570625
username                                              noitrecat
text          this war devil animation is magnificent \n#cha...
created_at                            2022-08-07 22:53:43+00:00
location                                                   None
source                                      Twitter for Android
url           https://twitter.com/noitrecat/status/155641330...
Name: 0, dtype: object

In [24]:
wardevil_df.head()

tweet_id      username  \
0  1557411045563564038  Bee_Curious_   
0  1557401799866388480   BERSERKISMS   
0  1557396517128110080  Ilumix_Naomi   
0  1557390855362826240    angeIkiIls   
0  1557366178275893248       dnbooru   

                                                text  \
0  Please like and follow! Chainsaw man's new cha...   
0  / it seems there are More war devil accs then ...   
0              War devil WIP https://t.co/NdsGzQlE7b   
0  imagine if the way the war devil transformed w...   
0  World\n#chainsaw_man\nCharacters\n#makima_chai...   

                 created_at                      location  \
0 2022-08-10 16:58:23+00:00                          None   
0 2022-08-10 16:21:38+00:00                          𝐌𝐃𝐍𝐈   
0 2022-08-10 16:00:39+00:00                          None   
0 2022-08-10 15:38:09+00:00  22. he/they. nsfw sometimes!   
0 2022-08-10 14:00:06+00:00                         Tokyo   

                source                                                url  
0  Twitter for Android  https://twitter.com/Bee_Curious_/status/155741...  
0   Twitter for iPhone  https://twitter.com/BERSERKISMS/status/1557401...  
0  Twitter for Android  https://twitter.com/Ilumix_Naomi/status/155739...  
0   Twitter for iPhone  https://twitter.com/angeIkiIls/status/15573908...  
0         Danbooru Bot  https://twitter.com/dnbooru/status/15573661782...

In [25]:
# merged and final dataframe
frames = [mitaka_df, asamitaka_df, mitakaasa_df, wardevil_df]
tw_mitaka_df = pd.concat(frames)
tw_mitaka_df.head()

#tw_mitaka_df = tw_mitaka_df[~tw_mitaka_df.index.duplicated(keep='first')] #Removing dup indices later

tweet_id       username  \
0  1556927525313323009       denpochi   
0  1556917447805898754       Lizsa___   
0  1556905368520998913  politicfemboy   
0  1556888261964214272    killmekirby   
0  1556884584977006592  retromanjapan   

                                                text  \
0  hey, could u stop posting about mitaka? i ship...   
0  join friend's style challenge\nAsa Mitaka from...   
0  I desire the same as Asa Mitaka... https://t.c...   
0  why does my asa mitaka drawing look like yumek...   
0  The Showa Retro In #MITAKA #Tokyo #8: A One-St...   

                 created_at                                         location  \
0 2022-08-09 08:57:03+00:00  17 they/them\ncsm bllk jjk opm\nic kamonegioisi   
0 2022-08-09 08:17:00+00:00                                             None   
0 2022-08-09 07:29:00+00:00                                             None   
0 2022-08-09 06:21:01+00:00                                     she/her • 17   
0 2022-08-09 06:06:25+00:00                                     Tokyo, Japan   

                source                                                url  
0  Twitter for Android  https://twitter.com/denpochi/status/1556927525...  
0  Twitter for Android  https://twitter.com/Lizsa___/status/1556917447...  
0      Twitter Web App  https://twitter.com/politicfemboy/status/15569...  
0   Twitter for iPhone  https://twitter.com/killmekirby/status/1556888...  
0    Revive Social App  https://twitter.com/retromanjapan/status/15568...

In [26]:
tw_mitaka_df.nunique()

tweet_id      346
username      276
text          334
created_at    345
location      144
source         16
url           346
dtype: int64

In [27]:
tw_mitaka_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 443 entries, 0 to 0
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype              
---  ------      --------------  -----              
 0   tweet_id    443 non-null    int64              
 1   username    443 non-null    object             
 2   text        443 non-null    object             
 3   created_at  443 non-null    datetime64[ns, UTC]
 4   location    232 non-null    object             
 5   source      443 non-null    object             
 6   url         443 non-null    object             
dtypes: datetime64[ns, UTC](1), int64(1), object(5)
memory usage: 27.7+ KB


In [28]:
tw_mitaka_df.to_csv('tw_mitaka_a.csv', index=False)

Since I can only query from the past 7 days, need to do more queries in future to get more datapoints.

8/5 Update: 846 unique tweets from 7/29 - 8/7 \
8/12 Update: 346 unique tweets from 8/7 - 8/14